lib/src/infrastructure/selectable_text.dart

```diff
@@ -0,0 +1,903 @@
+    if (_renderParagraph == null) {
+      return TextSelection.collapsed(offset: -1);
+    }
+
+    final wordRange = _renderParagraph!.getWordBoundary(position);
+    return TextSelection(
+      baseOffset: wordRange.start,
+      extentOffset: wordRange.end,
+    );
+  }
+
+  bool isTextAtOffset(Offset localOffset) {
+    if (_renderParagraph == null) {
+      return false;
+    }
+
+    List<TextBox> boxes = _renderParagraph!.getBoxesForSelection(
+      TextSelection(
+        baseOffset: 0,
+        extentOffset: _textLength,
+      ),
+    );
+
+    for (final box in boxes) {
+      if (box.toRect().contains(localOffset)) {
+        return true;
+      }
+    }
+
+    return false;
+  }
+
+  @override
+  Rect calculateLocalOverlap({
+    required Rect region,
+    required RenderObject ancestorCoordinateSpace,
+  }) {
+    if (_renderParagraph == null) {
+      return Rect.zero;
+    }
+
+    final renderParagraph = _renderParagraph!;
+    final contentOffset = renderParagraph.localToGlobal(Offset.zero, ancestor: ancestorCoordinateSpace);
+    final textRect = contentOffset & renderParagraph.size;
+
+    if (region.overlaps(textRect)) {
+      // Report the overlap in our local coordinate space.
+      return region.translate(-contentOffset.dx, -contentOffset.dy);
+    } else {
+      return Rect.zero;
+    }
+  }
+
+  @override
+  TextSelection getSelectionInRect(Offset baseOffset, Offset extentOffset) {
+    if (_renderParagraph == null) {
+      return TextSelection.collapsed(offset: -1);
+    }
+
+    final renderParagraph = _renderParagraph!;
+    final contentHeight = renderParagraph.size.height;
+    final textLength = _textLength;
+
+    // We don't know whether the base offset is higher or lower than the
+    // extent offset. Regardless, if either offset is above the top of
+    // the text then that text position should be 0. If either offset
+    // is below the bottom of the text then that offset should be the
+    // total length of the text.
+    final basePosition = baseOffset.dy < 0
+        ? 0
+        : baseOffset.dy > contentHeight
+            ? textLength
+            : renderParagraph.getPositionForOffset(baseOffset).offset;
+    final extentPosition = extentOffset.dy < 0
+        ? 0
+        : extentOffset.dy > contentHeight
+            ? textLength
+            : renderParagraph.getPositionForOffset(extentOffset).offset;
+
+    final selection = TextSelection(
+      baseOffset: basePosition,
+      extentOffset: extentPosition,
+    );
+
+    return selection;
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    if (_renderParagraph == null) {
+      WidgetsBinding.instance!.addPostFrameCallback((timeStamp) {
+        setState(() {
+          // Force another frame so that we can use the renderParagraph.
+        });
+      });
+    }
+
+    return Stack(
+      children: [
+        _buildTextSelection(),
+        _buildText(),
+        _buildTextCaret(),
+      ],
+    );
```

Comment:
We could avoid the `if (_renderParagraph == null) return const SizedBox();` blocks in the following methods with:
```suggestion
    return Stack(
      children: [
        if (_renderParagraph == null) _buildTextSelection(),
        _buildText(),
        if (_renderParagraph == null) _buildTextCaret(),
      ],
    );
```

Reply:
Looks like a tradeoff either way, I think I'd opt to keep the primary tree simpler (without the if-statements) and then allow the other build methods to decide what to return.